fix(docs-infra): preserve focus on copy (and prevent scrolling to bottom on IE11)

gkalpak opened the pull request:
The `CopierService` is used for copying text to the user's clipboard. It is, for example, used in `CodeComponent` to copy example code snippets. This is implemented by creating a temporary, hidden `<textarea>` elements, setting its value to the text that needs to be copied, executing the `copy` command and finally removing the element from the DOM.

Previously, as a result of `CopierService`'s implementation, the focused element would lose focus, while the temporary `<textarea>` element would implicitly gain focus when selecting its contents. This had an even worse side-effect on IE11, which seems to scroll to the bottom of the containing element (here `<body>`) when the focused element is removed.

This PR fixes these issues by keeping track of the previously focused element and restoring its focus after the copy operation. It also includes some minor refactorings. See individual commits for more details.

mary-poppins commented:
You can preview 84bd413 at https://pr38244-84bd413.ngbuilds.io/.

saifabusaleh commented on aio/src/app/shared/copier.service.ts:
```diff
@@ -5,63 +5,87 @@
+   * @param text The text to be inserted into the textarea.
+   * @return The temporary `<textarea>` element containing the specified text.
+   */
+  private createTextArea(text: string): HTMLTextAreaElement {
```
this function does two things, 
it creates text area
it adds it to DOM

Is this name is good enough?

gkalpak replied:
Good catch. I will update the function to not add the testarea to the DOM. I'll move the operqtion to the `copyText()` method, which is also responsible for removing the element from the DOM (so making it responsible for also inserting it makes sense).

saifabusaleh commented:
Hi @gkalpak ,

I wrote few comments

in MDN is saying that `select` is not always doing `focus`:
https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/select


but anyway returning the `focus` to previously focused element sounds like good fix 👍 

LGTM 

Thanks

saifabusaleh commented on aio/src/app/shared/copier.service.ts:
```diff
@@ -5,63 +5,87 @@
+    } catch {
+      return false;
+    } finally {
+      currentFocus?.focus();
```
Maybe its needed to write comment why this is needed after the `select` ?

gkalpak replied:
Good call. I'll add a comment.

saifabusaleh commented:
I see also now that the white border is now persistent after clicking the icon to copy, while it wasn't like that before

![image](https://user-images.githubusercontent.com/15147809/88554152-d347ea00-d02e-11ea-9236-8cb76ae45650.png)

gkalpak commented:
Thank you for the review, @saifabusaleh ❤️ I have made the changes.

> I see also now that the white border is now persistent after clicking the icon to copy, while it wasn't like that before

This is now happening because the `copy` button retains its focus (and what you see is its focus style). Previously, the focus would go to the hidden textarea, so the `copy` button would never be focused.

If we don't like the focus style, we can improve it in a separate PR (but such a change would be outside the scope of this PR).

mary-poppins commented:
You can preview e49b49d at https://pr38244-e49b49d.ngbuilds.io/.

saifabusaleh commented:
@gkalpak LGTM

Thanks 👍

kapunahelewong commented on aio/src/app/shared/copier.service.ts:
```diff
@@ -5,63 +5,95 @@
 export class CopierService {
-    private fakeElem: HTMLTextAreaElement|null;
+  /**
+   * Copy the contents of a `<textarea>` element to clipboard.
```
```suggestion
   * Copy the contents of a `<textarea>` element to the clipboard.
```

kapunahelewong commented on aio/src/app/shared/copier.service.ts:
```diff
@@ -5,63 +5,95 @@
+   *
+   * NOTE: For this method to work, the elements must be already inserted into the DOM.
+   *
+   * @param textArea The area containing the text to be copied to clipboard.
```
```suggestion
   * @param textArea The area containing the text to be copied to the clipboard.
```

kapunahelewong commented on aio/src/app/shared/copier.service.ts:
```diff
@@ -5,63 +5,95 @@
-      }
-    }
+  /**
+   * Copy the specified text to clipboard.
```
```suggestion
   * Copy the specified text to the clipboard.
```

kapunahelewong commented on aio/src/app/shared/copier.service.ts:
```diff
@@ -5,63 +5,95 @@
+  /**
+   * Copy the specified text to clipboard.
+   *
+   * @param text The text to be copied to clipboard.
```
```suggestion
   * @param text The text to be copied to the clipboard.
```

kapunahelewong reviewed (approved):
LGTM! (just a few minor notes)

Reviewed-for: global-docs-approvers

gkalpak committed:
fixup! refactor(docs-infra): improve code readability of `CopierService`

gkalpak committed:
refactor(docs-infra): improve code readability of `CopierService`

This commit improves the code readability of the `CopierService` by:
- Adding/Improving JSDoc comments for methods.
- Avoiding unnecessary instance-wide properties.
- Fixing indentation to be consistent (at two spaces).
- Clearly separating the logic for creating and populating a
  `<textarea>` from the logic for selecting and copying its contents.

gkalpak committed:
fix(docs-infra): preserve focus on copy (and prevent scrolling to bottom on IE11)

The `CopierService` is used for copying text to the user's clipboard. It
is, for example, used in `CodeComponent` to copy example code snippets.
This is implemented by creating a temporary, hidden `<textarea>`
elements, setting its value to the text that needs to be copied,
executing the `copy` command and finally removing the element from the
DOM.

Previously, as a result of `CopierService`'s implementation, the focused
element would lose focus, while the temporary `<textarea>` element would
implicitly gain focus when selecting its contents. This had an even
worse side-effect on IE11, which seems to scroll to the bottom of the
containing element (here `<body>`) when the focused element is removed.

This commit fixes these issues by keeping track of the previously
focused element and restoring its focus after the copy operation.

NOTE: This fix is inspired by Angular CDK's [PendingCopy][1] class.

[1]: https://github.com/angular/components/blob/89b5fa89d1437c3054c5/src/cdk/clipboard/pending-copy.ts

Fixes #37796

gkalpak committed:
refactor(docs-infra): simplify/improve `CopierService` hidden textarea creation

This commit simplifies the creation of the temporary, hidden
`<textarea>` element used by `CopierService` by switching from absolute
to fixed positioning and not requiring page's scroll offset.

It also makes the following minor improvements:
- Make the element invisible (via `opacity: 0`).
- Instruct screen-readers to ignore the element (via
  `aria-hidden: true`).

NOTE: These improvements are based on Angular CDK's [PendingCopy][1]
      class and the changes proposed in PR angular/components#20073.

[1]: https://github.com/angular/components/blob/89b5fa89d1437c3054c5/src/cdk/clipboard/pending-copy.ts

gkalpak committed:
fixup! fix(docs-infra): preserve focus on copy (and prevent scrolling to bottom on IE11)

gkalpak committed:
fixup! refactor(docs-infra): improve code readability of `CopierService`

mary-poppins commented:
You can preview 7260a56 at https://pr38244-7260a56.ngbuilds.io/.
